sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala

```diff
@@ -72,137 +74,191 @@ class OrcDeserializer(
 
         var i = 0
         val it = orcMap.entrySet().iterator()
         while (it.hasNext) {
           val entry = it.next()
-          keyConverter(i, entry.getKey)
+          keyConverter(keyUpdater, i, entry.getKey)
           val value = entry.getValue
           if (value == null) {
             valueArray.setNullAt(i)
           } else {
-            valueConverter(i, value)
+            valueConverter(valueUpdater, i, value)
           }
           i += 1
         }
 
         // The ORC map will never have null or duplicated map keys, it's safe to create a
         // ArrayBasedMapData directly here.
         updater.set(ordinal, new ArrayBasedMapData(keyArray, valueArray))
+      }
 
-      case udt: UserDefinedType[_] => newWriter(udt.sqlType, updater)
+    case udt: UserDefinedType[_] =>
+      val converter = newWriter(udt.sqlType, reuseObj)
+      (updater, ordinal, value) => {
+        converter(updater, ordinal, value)
+      }
+
+    case _ =>
+      throw new UnsupportedOperationException(s"$dataType is not supported yet.")
+  }
+
+  private def getArrayDataCreator(elementType: DataType): Int => ArrayData = elementType match {
+    case BooleanType => length => UnsafeArrayData.createFreshArray(length, 1)
+    case ByteType => length => UnsafeArrayData.createFreshArray(length, 1)
+    case ShortType => length => UnsafeArrayData.createFreshArray(length, 2)
+    case IntegerType => length => UnsafeArrayData.createFreshArray(length, 4)
+    case LongType => length => UnsafeArrayData.createFreshArray(length, 8)
+    case FloatType => length => UnsafeArrayData.createFreshArray(length, 4)
+    case DoubleType => length => UnsafeArrayData.createFreshArray(length, 8)
+    case _ => length => new GenericArrayData(new Array[Any](length))
+  }
 
-      case _ =>
-        throw new UnsupportedOperationException(s"$dataType is not supported yet.")
+  private def getRowCreator(st: StructType): () => InternalRow = {
+    val constructorsArray = new Array[() => MutableValue](st.fields.length)
+    var i = 0
+    while (i < st.fields.length) {
+      st.fields(i).dataType match {
+        case BooleanType => constructorsArray(i) = () => new MutableBoolean
+        case ByteType => constructorsArray(i) = () => new MutableByte
+        case ShortType => constructorsArray(i) = () => new MutableShort
+        // We use INT for DATE internally
+        case IntegerType | DateType => constructorsArray(i) = () => new MutableInt
+        // We use Long for Timestamp internally
+        case LongType | TimestampType => constructorsArray(i) = () => new MutableLong
+        case FloatType => constructorsArray(i) = () => new MutableFloat
+        case DoubleType => constructorsArray(i) = () => new MutableDouble
+        case _ => constructorsArray(i) = () => new MutableAny
+      }
+      i += 1
     }
 
-  private def createArrayData(elementType: DataType, length: Int): ArrayData = elementType match {
```

Comment:
Here as well. If we can get the length from the type as you did `getArrayDataCreator`, looks we can just set it here and pull the `createArrayData` logic out of the function so it creates once. Presumably, same as `MapType`.

Reply:
Changed now.